partialmessages/partialmsgs.go

```diff
@@ -0,0 +1,290 @@
+package partialmessages
+
+import (
+	"bytes"
+	"errors"
+	"iter"
+	"log/slog"
+	"slices"
+
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// TODO: Add gossip fallback (pick random connected peers and send ihave/iwants)
+// TODO: Update scoring code to not kick partial-message only peers
+// Question: How to configure scheduling RPCs?
+//   - Leave it up to the implementation of sendRPC
+// Question: Skip partial IHAVE for now?
+// Question: I could have a user provided validation queue instead of requiring republishing
+//   - But a user may need to republish anyways if they get parts out of band
+
+const minGroupTTL = 3
+
+// PartialMessage is a message that can be broken up into parts. It can be
+// complete, partially complete, or empty. It is up to the application to define
+// how a message is split into parts and recombined, as well as how missing and
+// available parts are represented.
+//
+// It is passed to Gossipsub with a PublishPartialMessage method call. Gossipsub
+// keeps a reference to this object, so the implementation should either not
+// mutate this object in a separate goroutine after handing it to Gossipsub, or
+// take care to make the object thread safe.
+type PartialMessage interface {
```

Comment:
NIT: partialmessages is the package name, so we can remove Partial from the type names:

read by users as `partialmessages.Message`

```suggestion
type Message interface {
```